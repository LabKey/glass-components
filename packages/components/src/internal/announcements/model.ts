import { ActionURL, User } from '@labkey/api';

// Expected to match WikiRendererType.java
export enum AnnouncementRenderType {
    HTML = 'HTML',
    MARKDOWN = 'MARKDOWN',
    RADEOX = 'RADEOX',
    TEXT_WITH_LINKS = 'TEXT_WITH_LINKS',
}

export interface Attachment {
    created?: string;
    documentSize: number;
    name: string;
    parent?: string;
}

export const getAttachmentURL = (attachment: Attachment, containerPath?: string): string => {
    return ActionURL.buildURL('announcements', 'download', containerPath, {
        name: attachment.name,
        entityId: attachment.parent,
    });
};

export interface AnnouncementModel {
    approved?: string;
    author: User;
    attachments: Attachment[];
    body: string;
    containerId: string;
    containerPath: string;
    created: string;
    createdBy: number;
    discussionSrcIdentifier?: string;
    discussionSrcEntityType?: string;
    entityId: string;
    expires?: string;
    formattedHtml: string;
    modified: string;
    modifiedBy: number;
    parent?: string;
    rendererType: AnnouncementRenderType;
    responses: AnnouncementModel[];
    rowId: number;
    status?: string;
    title: string;
}

// See copyEditableProps from AnnouncementsController
export const EDITABLE_FIELDS = [
    'approved',
    'body',
    'discussionSrcIdentifier',
    'discussionSrcEntityType',
    'expires',
    'parent',
    'rendererType',
    'rowId',
    'status',
    'title',
];

export interface ThreadActions {
    onCreate: (newThread: AnnouncementModel) => void;
    onDelete: () => void;
    onUpdate: () => void;
}
