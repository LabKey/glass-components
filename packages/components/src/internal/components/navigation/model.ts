/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { List, Record } from 'immutable';
import { Ajax, Utils, QueryKey } from '@labkey/api';

import { buildURL, createProductUrl, AppURL, createProductUrlFromPartsWithContainer } from '../../url/AppURL';

export class MenuSectionModel extends Record({
    label: undefined,
    url: undefined,
    items: List<MenuItemModel>(),
    totalCount: 0,
    key: undefined,
    productId: undefined,
    sectionKey: undefined,
}) {
    declare label: string;
    declare url: string;
    declare items: List<MenuItemModel>;
    declare totalCount: number;
    declare key: string;
    declare productId: string;
    declare sectionKey: string;

    static create(rawData: any, currentProductId?: string, containerPath?: string): MenuSectionModel {
        if (rawData) {
            let items;

            if (rawData.items) {
                items = rawData.items.map(i =>
                    MenuItemModel.create(i, rawData.sectionKey, currentProductId, containerPath)
                );
            }

            return new MenuSectionModel(
                Object.assign({}, rawData, {
                    items: List(items),
                })
            );
        }

        return new MenuSectionModel();
    }

    getVisibleItems(): MenuItemModel[] {
        return this.items.filter(item => !item.hidden).toArray();
    }

    hasVisibleItems(): boolean {
        return this.getVisibleItems().length > 0;
    }
}

export class MenuItemModel extends Record({
    id: undefined,
    key: undefined,
    label: undefined,
    url: undefined,
    hidden: false,
    orderNum: undefined,
    originalUrl: undefined,
    requiresLogin: false,
    hasActiveJob: false,
    fromSharedContainer: false,
}) {
    declare id: number;
    declare key: string;
    declare label: string;
    declare url: string | AppURL;
    declare hidden: boolean;
    declare orderNum: number;
    declare originalUrl: string;
    declare requiresLogin: boolean;
    declare hasActiveJob: boolean;
    declare fromSharedContainer: boolean;

    static create(rawData, sectionKey: string, currentProductId?: string, containerPath?: string): MenuItemModel {
        if (rawData) {
            const dataProductId = rawData.productId ? rawData.productId.toLowerCase() : undefined;

            if (rawData.key && sectionKey !== 'user') {
                // for assay names and freezer names that contain slashes, full raw key (protocol/assayname: general/a/b) is encoded using QueryKey.encodePart as general/a$Sb on server side
                // use QueryKey.decodePart to decode the name so url can be correctly called by encodeURIComponent and key be used to display decoded assay name
                const subParts = rawData.key
                    .split('/')
                    .filter(val => val !== '')
                    .map(QueryKey.decodePart);

                const decoded = subParts.join('/');
                const decodedKey = rawData.key.replace(rawData.key, () => decoded); // use the functional version to skip any additional pattern substitutions https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace#specifying_a_string_as_the_replacement

                return new MenuItemModel(
                    Object.assign({}, rawData, {
                        originalUrl: rawData.url,
                        url: createProductUrlFromPartsWithContainer(
                            dataProductId,
                            currentProductId,
                            containerPath,
                            undefined,
                            sectionKey,
                            ...subParts
                        ),
                        key: decodedKey,
                    })
                );
            } else {
                return new MenuItemModel(
                    Object.assign({}, rawData, {
                        originalUrl: rawData.url,
                        url: createProductUrl(dataProductId, currentProductId, rawData.url, containerPath),
                    })
                );
            }
        }
        return new MenuItemModel();
    }

    getUrlString(useOriginalUrl?: boolean): string {
        if (useOriginalUrl && this.originalUrl) return this.originalUrl;
        return typeof this.url === 'string' ? this.url : this.url?.toHref();
    }
}

export class ProductMenuModel extends Record({
    isError: false,
    isLoaded: false,
    isLoading: false,
    sections: List<MenuSectionModel>(),
    message: undefined,
    containerId: undefined,
    containerPath: undefined,
    currentProductId: undefined,
    productIds: undefined,
    needsReload: false,
}) {
    declare isError: boolean;
    declare isLoaded: boolean;
    declare isLoading: boolean;
    declare message: string;
    declare sections: List<MenuSectionModel>;
    declare containerId: string;
    declare containerPath: string;
    declare currentProductId: string; // the current product's id
    declare productIds: List<string>; // the list of all product ids to be included in the menu; leave undefined for all products in the container
    declare needsReload: boolean;

    /**
     * Retrieve the product menu sections for this productId
     */
    getMenuSections(): Promise<List<MenuSectionModel>> {
        return new Promise((resolve, reject) => {
            return Ajax.request({
                url: buildURL('product', 'menuSections.api', undefined, {
                    container: this.containerPath,
                }),
                params: Object.assign({
                    currentProductId: this.currentProductId,
                    productIds: List.isList(this.productIds) ? this.productIds.toArray().join(',') : this.productIds,
                }),
                success: Utils.getCallbackWrapper(response => {
                    let sections = List<MenuSectionModel>();
                    if (response) {
                        response.forEach(sectionData => {
                            sections = sections.push(
                                MenuSectionModel.create(sectionData, this.currentProductId, this.containerPath)
                            );
                        });
                    }
                    resolve(sections);
                }),
                failure: Utils.getCallbackWrapper(response => {
                    console.error(response);
                    reject(response);
                }),
            });
        });
    }

    setLoadedSections(sections: List<MenuSectionModel>): ProductMenuModel {
        return this.merge({
            isLoaded: true,
            isLoading: false,
            needsReload: false,
            sections,
        }) as ProductMenuModel;
    }

    setError(message: string): ProductMenuModel {
        return this.merge({
            isLoaded: true,
            isLoading: false,
            needsReload: false,
            isError: true,
            message,
        }) as ProductMenuModel;
    }

    getSection(key: string): MenuSectionModel {
        return this.sections.find(section => section.key.toLowerCase() === key.toLowerCase());
    }

    hasSectionItems(key: string): boolean {
        return this.isLoaded && this.getSection(key)?.totalCount > 0;
    }

    setNeedsReload(): ProductMenuModel {
        return this.merge({
            needsReload: true,
        }) as ProductMenuModel;
    }
}

export class MenuSectionConfig extends Record({
    activeJobIconCls: 'fa-spinner fa-pulse',
    emptyText: undefined,
    filteredEmptyText: undefined,
    emptyAppURL: undefined,
    emptyURLText: 'Get started...',
    headerURLPart: undefined,
    headerText: undefined,
    iconCls: undefined,
    iconURL: undefined,
    showActiveJobIcon: true,
    staticContent: false,
    useOriginalURL: false,
}) {
    declare activeJobIconCls?: string;
    declare emptyText?: string;
    declare filteredEmptyText?: string;
    declare emptyAppURL?: AppURL;
    declare emptyURLText: string;
    declare headerURLPart: string;
    declare headerText?: string;
    declare iconCls?: string;
    declare iconURL?: string;
    declare showActiveJobIcon?: boolean;
    // Inform the display that this section's content is static (unchanging).
    // This helps inform the layout when these sections are laid out alongside sections with dynamic content.
    declare staticContent?: boolean;
    declare useOriginalURL?: boolean;
}
