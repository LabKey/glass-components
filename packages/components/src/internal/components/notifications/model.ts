/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { Record } from 'immutable';

import { generateId, User } from '../../..';
import { Draft, immerable, produce } from "immer";

export type MessageFunction<T> = (props?: T, user?: User, data?: any) => React.ReactNode;

export const enum Persistence {
    PAGE_LOAD,
    LOGIN_SESSION,
}

// This interface is available so you do not have .create a Model object and so the id can be optionally
// created for you.
export interface NotificationItemProps {
    alertClass?: string;
    data?: any;
    id?: string;
    isDismissible?: boolean;
    isDismissed?: boolean;
    message: string | MessageFunction<NotificationItemProps>;
    onDismiss?: () => any;
    persistence?: Persistence;
}

export class NotificationItemModel
    extends Record({
        alertClass: 'success',
        data: undefined,
        id: undefined,
        isDismissible: true,
        isDismissed: false,
        message: undefined,
        onDismiss: undefined,
        persistence: Persistence.PAGE_LOAD,
    })
    implements NotificationItemProps {
    alertClass: string;
    data?: any;
    id: string;
    isDismissible: boolean;
    isDismissed: boolean;
    message: string | MessageFunction<NotificationItemProps>;
    onDismiss?: () => any;
    persistence?: Persistence;

    static create(values?: NotificationItemProps): NotificationItemModel {
        return new NotificationItemModel(
            Object.assign(
                {},
                {
                    id: nextNotificationId(),
                },
                values
            )
        );
    }
}

function nextNotificationId(): string {
    return generateId('notification_');
}


export class ServerActivityData {
    [immerable] = true;

    readonly RowId: number;
    readonly Type: string;
    readonly Created: string;
    readonly CreatedBy: string;
    readonly UserId: number;
    readonly ObjectId: string;
    readonly ReadOn: string;
    readonly ActionLinkText: string;
    readonly ActionLinkUrl: string;
    readonly ContainerId: string;
    readonly HtmlContent: string;
    readonly ContentType: string;
    readonly IconCls: string;
    readonly inProgress: boolean;
    readonly hasError: boolean;

    constructor(values?: Partial<ServerActivityData>) {
        const addedValues = {};
        if (values.Type?.indexOf('error') >= 0) {
            Object.assign(addedValues, { hasError: true });
        }
        Object.assign(this, values, addedValues);
    }

    mutate(props: Partial<ServerActivityData>): ServerActivityData {
        return produce(this, (draft: Draft<ServerActivityData>) => {
            Object.assign(draft, props);
        });
    }

    isUnread(): boolean {
        return this.ReadOn == undefined;
    }
}

export interface ServerActivity {
    data: ServerActivityData[];
    totalRows: number;
    unreadCount: number;
    inProgressCount: number;
}

export interface ServerNotificationsConfig {
    maxRows: number;
    markAllNotificationsRead: () => Promise<boolean>;
    getNotificationData: (maxRows: number) => Promise<ServerActivity>;
}
