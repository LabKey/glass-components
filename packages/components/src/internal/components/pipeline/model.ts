import {Draft, immerable, produce} from "immer";

export interface PipelineLogEntry {
    dateTime: string
    level: string
    lines: string
    multiline: boolean
    stackTrace: boolean
}

export class PipelineStatusDetailModel {
    [immerable] = true;

    readonly rowId: number;
    readonly created: string;
    readonly status: string;
    readonly description: string;
    readonly info: string;
    readonly active: boolean;
    readonly fetchCount: number;
    readonly nextOffset: number;
    readonly logEntries: PipelineLogEntry[];

    readonly isLoaded: boolean;
    readonly isLoading: boolean;

    constructor(values?: Partial<PipelineStatusDetailModel>) {
        Object.assign(this, {isLoaded: false, isLoading: false, fetchCount: 1, nextOffset: 0}, values);
    }

    static loadResult(data?: any) {
        return new PipelineStatusDetailModel({
            rowId: data['rowId'],
            created: data['created'],
            status: data['status'],
            description: data['description'],
            info: data['info'],
            active: data['active'],
            fetchCount: data['fetchCount'],
            nextOffset: data['nextOffset'],
            logEntries: data['log'] ? data['log']['records'] : [],
            isLoaded: true,
            isLoading: false
        })
    }

    mutate(props: Partial<PipelineStatusDetailModel>): PipelineStatusDetailModel {
        return produce(this, (draft: Draft<PipelineStatusDetailModel>) => {
            Object.assign(draft, props);
        });
    }


}