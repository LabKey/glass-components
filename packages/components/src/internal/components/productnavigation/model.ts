import { immerable } from 'immer';

export class ProductModel {
    [immerable] = true;

    moduleName: string;
    sectionNames: string[];
    documentationUrl: string;
    productId: string;
    productName: string;
    documentationLabel: string;

    constructor(values?: Partial<ProductModel>) {
        Object.assign(this, values);
    }
}
