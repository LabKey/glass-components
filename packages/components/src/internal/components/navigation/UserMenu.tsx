/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { Dropdown, Image, MenuItem } from 'react-bootstrap';

import { User } from '../../..';
import { devToolsActive, toggleDevTools } from '../../util/utils';
import { buildURL } from '../../url/ActionURL';
import { AppURL } from '../../url/AppURL';

import { ProductMenuModel } from './model';
import { signOut, signIn } from './actions';

interface UserMenuProps {
    model: ProductMenuModel;
    user: User;
    showSwitchToLabKey: boolean;
    extraDevItems?: any;
    extraUserItems?: any;
    signOutUrl?: string;
}

export class UserMenu extends React.Component<UserMenuProps, any> {
    render() {
        const { extraDevItems, extraUserItems, model, user, showSwitchToLabKey, signOutUrl } = this.props;
        const menuSection = model.getSection('user');

        if (menuSection) {
            const beginUrl = buildURL('project', 'begin', undefined, { returnURL: false });
            const switchToLabKeyItem = (
                <MenuItem key="projectBegin" href={beginUrl}>
                    Switch to LabKey
                </MenuItem>
            );

            const menuItems = [];
            menuSection.items.forEach(item => {
                if ((item.requiresLogin && user.isSignedIn) || !item.requiresLogin) {
                    const href = item.url instanceof AppURL ? item.url.toHref() : item.url;
                    menuItems.push(
                        <MenuItem key={item.key} href={href} target={item.key === 'docs' ? '_blank' : '_self'}>
                            {item.label}
                        </MenuItem>
                    );
                }
            });

            return (
                <Dropdown id="user-menu-dropdown">
                    <Dropdown.Toggle useAnchor={true}>
                        {user.avatar ? (
                            <Image src={user.avatar} alt="User Avatar" rounded={true} height={32} width={32} />
                        ) : (
                            <span className="navbar-item">
                                <span className="user-name">
                                    <span className="fas fa-user-circle" /> {user.displayName}{' '}
                                </span>
                            </span>
                        )}
                    </Dropdown.Toggle>

                    <Dropdown.Menu pullRight className="pull-right">
                        <div className="navbar-connector" />
                        {menuItems}
                        {showSwitchToLabKey && switchToLabKeyItem}
                        {extraUserItems}
                        {LABKEY.devMode ? (
                            <>
                                <MenuItem divider />
                                <MenuItem header>Dev Tools</MenuItem>
                                <MenuItem onClick={toggleDevTools}>
                                    {devToolsActive() ? 'Disable' : 'Enable'} Redux Tools
                                </MenuItem>
                                {!showSwitchToLabKey && switchToLabKeyItem}
                                {extraDevItems}
                            </>
                        ) : null}
                        <MenuItem divider />
                        {user.isSignedIn ? (
                            <MenuItem onClick={() => signOut(signOutUrl)}>Sign Out</MenuItem>
                        ) : (
                            <MenuItem onClick={signIn}>Sign In</MenuItem>
                        )}
                    </Dropdown.Menu>
                </Dropdown>
            );
        }
        return null;
    }
}
