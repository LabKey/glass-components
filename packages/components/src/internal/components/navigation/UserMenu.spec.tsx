/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { List } from 'immutable';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import { User } from '../../..';

import { UserMenu } from './UserMenu';
import { MenuSectionModel, ProductMenuModel } from './model';

beforeAll(() => {
    LABKEY.devMode = false;
});

describe('UserMenu', () => {
    const sections = List([
        MenuSectionModel.create({
            key: 'user',
            label: 'Your Items',
            url: undefined,
            items: [
                {
                    key: 'profile',
                    label: 'Profile',
                    url: 'profile/link/here',
                    requiresLogin: true,
                },
                {
                    key: 'docs',
                    label: 'Documentation',
                    url: 'http://show/me/the/docs',
                    requiresLogin: false,
                },
            ],
            sectionKey: 'user',
        }),
    ]);

    test('not initialized', () => {
        const model = new ProductMenuModel({
            productIds: ['testProduct'],
        });
        const tree = mount(<UserMenu model={model} user={new User()} showSwitchToLabKey />);
        expect(tree).toEqual({});
    });

    test('user not logged in', () => {
        const productIds = ['notLoggedInUser'];
        const user = new User({
            isSignedIn: false,
        });

        const model = new ProductMenuModel({
            isLoaded: true,
            isLoading: false,
            productIds,
            sections,
        });
        const tree = renderer.create(<UserMenu model={model} user={user} showSwitchToLabKey />);
        expect(tree).toMatchSnapshot();
    });

    test('user logged in, but not in dev mode', () => {
        const productIds = ['loggedInUser'];
        const user = new User({
            isSignedIn: true,
        });

        const model = new ProductMenuModel({
            isLoaded: true,
            isLoading: false,
            productIds,
            sections,
        });
        const tree = renderer.create(<UserMenu model={model} user={user} showSwitchToLabKey />);
        expect(tree).toMatchSnapshot();
    });

    test('user logged in dev mode', () => {
        const productIds = ['logginedInDevMode'];
        const user = new User({
            isSignedIn: true,
        });
        LABKEY.devMode = true;
        const model = new ProductMenuModel({
            isLoaded: true,
            isLoading: false,
            productIds,
            sections,
        });
        const tree = renderer.create(<UserMenu model={model} user={user} showSwitchToLabKey />);
        expect(tree).toMatchSnapshot();
    });

    test('user logged in extra items', () => {
        const productIds = ['extraUserItems'];
        const user = new User({
            isSignedIn: true,
        });

        const model = new ProductMenuModel({
            isLoaded: true,
            isLoading: false,
            productIds,
            sections,
        });
        const extraUserItems = [<div key="e1">Extra One</div>, <div key="e2">Extra Two</div>];
        const tree = renderer.create(
            <UserMenu model={model} user={user} showSwitchToLabKey extraUserItems={extraUserItems} />
        );
        expect(tree).toMatchSnapshot();
    });

    test('user logged in, without switch to labkey', () => {
        const productIds = ['switchToLabkey'];
        const user = new User({
            isSignedIn: true,
        });

        const model = new ProductMenuModel({
            isLoaded: true,
            isLoading: false,
            productIds,
            sections,
        });
        const tree = renderer.create(<UserMenu model={model} user={user} showSwitchToLabKey={false} />);
        expect(tree).toMatchSnapshot();
    });

    test('user logged in extra dev mode items', () => {
        const productIds = ['extraDevItems'];
        const user = new User({
            isSignedIn: true,
        });

        const model = new ProductMenuModel({
            isLoaded: true,
            isLoading: false,
            productIds,
            sections,
        });
        const extraUserItems = [<div key="e1">Extra One</div>, <div key="e2">Extra Two</div>];
        const extraDevItems = [<div key="e1">Extra Dev One</div>, <div key="e2">Extra Dev Two</div>];
        const tree = renderer.create(
            <UserMenu
                extraDevItems={extraDevItems}
                extraUserItems={extraUserItems}
                model={model}
                showSwitchToLabKey
                user={user}
            />
        );
        expect(tree).toMatchSnapshot();
    });
});
