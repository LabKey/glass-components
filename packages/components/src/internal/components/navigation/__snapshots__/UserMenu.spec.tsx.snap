// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserMenu user logged in dev mode 1`] = `
<div
  className="dropdown btn-group"
>
  <a
    aria-expanded={false}
    aria-haspopup={true}
    className="dropdown-toggle"
    href="#"
    id="user-menu-dropdown"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
  >
    <img
      alt="User Avatar"
      className="img-rounded"
      height={32}
      src="/labkey/_images/defaultavatar.png"
      width={32}
    />
     
    <span
      className="caret"
    />
  </a>
  <ul
    aria-labelledby="user-menu-dropdown"
    className="pull-right dropdown-menu"
    role="menu"
  >
    <div
      className="navbar-connector"
      onKeyDown={[Function]}
      onSelect={[Function]}
    />
    <li
      className=""
      role="presentation"
    >
      <a
        href="profile/link/here"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
        target="_self"
      >
        Profile
      </a>
    </li>
    <li
      className=""
      role="presentation"
    >
      <a
        href="http://show/me/the/docs"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
        target="_blank"
      >
        Documentation
      </a>
    </li>
    <li
      className="divider"
      role="separator"
    />
    <li
      className="submenu-header dropdown-header"
      role="heading"
    >
      Dev Tools
    </li>
    <li
      className=""
      role="presentation"
    >
      <a
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
      >
        Enable
         Redux Tools
      </a>
    </li>
    <li
      className="divider"
      onKeyDown={[Function]}
      role="separator"
    />
    <li
      className=""
      role="presentation"
    >
      <a
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
      >
        Sign Out
      </a>
    </li>
  </ul>
</div>
`;

exports[`UserMenu user logged in extra dev mode items 1`] = `
<div
  className="dropdown btn-group"
>
  <a
    aria-expanded={false}
    aria-haspopup={true}
    className="dropdown-toggle"
    href="#"
    id="user-menu-dropdown"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
  >
    <img
      alt="User Avatar"
      className="img-rounded"
      height={32}
      src="/labkey/_images/defaultavatar.png"
      width={32}
    />
     
    <span
      className="caret"
    />
  </a>
  <ul
    aria-labelledby="user-menu-dropdown"
    className="pull-right dropdown-menu"
    role="menu"
  >
    <div
      className="navbar-connector"
      onKeyDown={[Function]}
      onSelect={[Function]}
    />
    <li
      className=""
      role="presentation"
    >
      <a
        href="profile/link/here"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
        target="_self"
      >
        Profile
      </a>
    </li>
    <li
      className=""
      role="presentation"
    >
      <a
        href="http://show/me/the/docs"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
        target="_blank"
      >
        Documentation
      </a>
    </li>
    <div
      onKeyDown={[Function]}
      onSelect={[Function]}
    >
      Extra One
    </div>
    <div
      onKeyDown={[Function]}
      onSelect={[Function]}
    >
      Extra Two
    </div>
    <li
      className="divider"
      role="separator"
    />
    <li
      className="submenu-header dropdown-header"
      role="heading"
    >
      Dev Tools
    </li>
    <li
      className=""
      role="presentation"
    >
      <a
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
      >
        Enable
         Redux Tools
      </a>
    </li>
    <div>
      Extra Dev One
    </div>
    <div>
      Extra Dev Two
    </div>
    <li
      className="divider"
      onKeyDown={[Function]}
      role="separator"
    />
    <li
      className=""
      role="presentation"
    >
      <a
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
      >
        Sign Out
      </a>
    </li>
  </ul>
</div>
`;

exports[`UserMenu user logged in extra items 1`] = `
<div
  className="dropdown btn-group"
>
  <a
    aria-expanded={false}
    aria-haspopup={true}
    className="dropdown-toggle"
    href="#"
    id="user-menu-dropdown"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
  >
    <img
      alt="User Avatar"
      className="img-rounded"
      height={32}
      src="/labkey/_images/defaultavatar.png"
      width={32}
    />
     
    <span
      className="caret"
    />
  </a>
  <ul
    aria-labelledby="user-menu-dropdown"
    className="pull-right dropdown-menu"
    role="menu"
  >
    <div
      className="navbar-connector"
      onKeyDown={[Function]}
      onSelect={[Function]}
    />
    <li
      className=""
      role="presentation"
    >
      <a
        href="profile/link/here"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
        target="_self"
      >
        Profile
      </a>
    </li>
    <li
      className=""
      role="presentation"
    >
      <a
        href="http://show/me/the/docs"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
        target="_blank"
      >
        Documentation
      </a>
    </li>
    <div
      onKeyDown={[Function]}
      onSelect={[Function]}
    >
      Extra One
    </div>
    <div
      onKeyDown={[Function]}
      onSelect={[Function]}
    >
      Extra Two
    </div>
    <li
      className="divider"
      role="separator"
    />
    <li
      className="submenu-header dropdown-header"
      role="heading"
    >
      Dev Tools
    </li>
    <li
      className=""
      role="presentation"
    >
      <a
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
      >
        Enable
         Redux Tools
      </a>
    </li>
    <li
      className="divider"
      onKeyDown={[Function]}
      role="separator"
    />
    <li
      className=""
      role="presentation"
    >
      <a
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
      >
        Sign Out
      </a>
    </li>
  </ul>
</div>
`;

exports[`UserMenu user logged in, but not in dev mode 1`] = `
<div
  className="dropdown btn-group"
>
  <a
    aria-expanded={false}
    aria-haspopup={true}
    className="dropdown-toggle"
    href="#"
    id="user-menu-dropdown"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
  >
    <img
      alt="User Avatar"
      className="img-rounded"
      height={32}
      src="/labkey/_images/defaultavatar.png"
      width={32}
    />
     
    <span
      className="caret"
    />
  </a>
  <ul
    aria-labelledby="user-menu-dropdown"
    className="pull-right dropdown-menu"
    role="menu"
  >
    <div
      className="navbar-connector"
      onKeyDown={[Function]}
      onSelect={[Function]}
    />
    <li
      className=""
      role="presentation"
    >
      <a
        href="profile/link/here"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
        target="_self"
      >
        Profile
      </a>
    </li>
    <li
      className=""
      role="presentation"
    >
      <a
        href="http://show/me/the/docs"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
        target="_blank"
      >
        Documentation
      </a>
    </li>
    <li
      className="divider"
      onKeyDown={[Function]}
      role="separator"
    />
    <li
      className=""
      role="presentation"
    >
      <a
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
      >
        Sign Out
      </a>
    </li>
  </ul>
</div>
`;

exports[`UserMenu user not logged in 1`] = `
<div
  className="dropdown btn-group"
>
  <a
    aria-expanded={false}
    aria-haspopup={true}
    className="dropdown-toggle"
    href="#"
    id="user-menu-dropdown"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
  >
    <img
      alt="User Avatar"
      className="img-rounded"
      height={32}
      src="/labkey/_images/defaultavatar.png"
      width={32}
    />
     
    <span
      className="caret"
    />
  </a>
  <ul
    aria-labelledby="user-menu-dropdown"
    className="pull-right dropdown-menu"
    role="menu"
  >
    <div
      className="navbar-connector"
      onKeyDown={[Function]}
      onSelect={[Function]}
    />
    <li
      className=""
      role="presentation"
    >
      <a
        href="http://show/me/the/docs"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
        target="_blank"
      >
        Documentation
      </a>
    </li>
    <li
      className="divider"
      onKeyDown={[Function]}
      role="separator"
    />
    <li
      className=""
      role="presentation"
    >
      <a
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
      >
        Sign In
      </a>
    </li>
  </ul>
</div>
`;
